File: server/terrain/perlin.go
// Parts of this file were created by nsf on GitHub
// https://gist.github.com/nsf/1170424

package terrain

import (
    "log"
    "math"
    "math/rand"
)

const PI = 3.1415926535

type Vec2 struct {
    X, Y float64
}

func lerp(a, b, v float64) float64 {
    return a * (1 - v) + b * v
}

func smooth(v float64) float64 {
    return v * v * (3 - 2 * v)
}

func random_gradient(r *rand.Rand) Vec2 {
    v := r.Float64() * PI * 2
    return Vec2{
        float64(math.Cos(v)),
        float64(math.Sin(v)),
    }
}

func gradient(orig, grad, p Vec2) float64 {
    sp := Vec2{p.X - orig.X, p.Y - orig.Y}
    return grad.X * sp.X + grad.Y * sp.Y
}

type NoiseGenerator struct {
    rgradients []Vec2
    permutations []int
    gradients [4]Vec2
    origins [4]Vec2
}

func NewNoiseGenerator(seed int) *NoiseGenerator {
    log.Println("seed", seed)
    rng := rand.New(rand.NewSource(int64(seed)))

    gen := new(NoiseGenerator)
    gen.rgradients = make([]Vec2, TERRAIN_PERLIN_PERIOD)
    gen.permutations = rng.Perm(TERRAIN_PERLIN_PERIOD)
    for i := range gen.rgradients {
        gen.rgradients[i] = random_gradient(rng)
    }

    return gen
}

func (self *NoiseGenerator) get_gradient(x, y int) Vec2 {
    idx := self.permutations[x & (TERRAIN_PERLIN_PERIOD - 1)] + self.permutations[y & (TERRAIN_PERLIN_PERIOD - 1)]
    return self.rgradients[idx & (TERRAIN_PERLIN_PERIOD - 1)]
}

func (self *NoiseGenerator) get_gradients(x, y float64) {
    x0f := math.Floor(float64(x))
    y0f := math.Floor(float64(y))
    x0 := int(x0f)
    y0 := int(y0f)
    x1 := x0 + 1
    y1 := y0 + 1

    self.gradients[0] = self.get_gradient(x0, y0)
    self.gradients[1] = self.get_gradient(x1, y0)
    self.gradients[2] = self.get_gradient(x0, y1)
    self.gradients[3] = self.get_gradient(x1, y1)

    self.origins[0] = Vec2{float64(x0f + 0.0), float64(y0f + 0.0)}
    self.origins[1] = Vec2{float64(x0f + 1.0), float64(y0f + 0.0)}
    self.origins[2] = Vec2{float64(x0f + 0.0), float64(y0f + 1.0)}
    self.origins[3] = Vec2{float64(x0f + 1.0), float64(y0f + 1.0)}
}

func (self *NoiseGenerator) Get2D(x, y float64) float64 {
    p := Vec2{x, y}
    self.get_gradients(x, y)
    v0 := gradient(self.origins[0], self.gradients[0], p)
    v1 := gradient(self.origins[1], self.gradients[1], p)
    v2 := gradient(self.origins[2], self.gradients[2], p)
    v3 := gradient(self.origins[3], self.gradients[3], p)
    fx := smooth(x - self.origins[0].X)
    vx0 := lerp(v0, v1, fx)
    vx1 := lerp(v2, v3, fx)
    fy := smooth(y - self.origins[0].Y)
    return lerp(vx0, vx1, fy)
}

func (self *NoiseGenerator) Get2DInt(x, y int, max uint) int {
    v := self.Get2D(float64(x) * PERLIN_FREQUENCY, float64(y) * PERLIN_FREQUENCY)
    v = v * 0.5 + 0.5
    v = math.Pow(v, PERLIN_DILATION)
    return int(v * float64(max))
}

func (self *NoiseGenerator) GetCentered2DInt(x, y int, freq float64, bounds uint) int {
    v := self.Get2D(float64(x) * freq, float64(y) * freq)
    // v = math.Pow(v, PERLIN_DILATION)
    return int(v * float64(bounds))
}

func (self *NoiseGenerator) FillGrid(x, y int, grid *[][]uint, max uint) {
    // log.Println("Perlin for ", x, y)
    // log.Println(self.Get2D(0 * PERLIN_FREQUENCY, 0 * PERLIN_FREQUENCY) * 100)
    // log.Println(self.Get2D(float64(x) * PERLIN_FREQUENCY, float64(y) * PERLIN_FREQUENCY) * 100)
    for i := 0; i < len(*grid); i++ {
        for j := 0; j < len((*grid)[i]); j++ {
            (*grid)[i][j] = uint(
                // self.Get2DInt(x + j, y + i, max - 1) +
                self.GetCentered2DInt(x + j, y + i, PERLIN_BIOME_FREQUENCY, max - 1) +
                PERLIN_UPLIFT,
            )
        }
    }
}

File: src/terrainGen/perlin.js
const rng = require('rng');


const TERRAIN_PERLIN_PERIOD     = 256;
const TERRAIN_PERLIN_PERIOD_M1  = TERRAIN_PERLIN_PERIOD - 1;
const TERRAIN_PERLIN_MAX        = 6;
const PERLIN_FREQUENCY          = 0.1;
const PERLIN_DILATION           = 3;
const PERLIN_BIOME_FREQUENCY    = 0.0075;
const PERLIN_UPLIFT             = 5;


function lerp(a, b, v) {
  return a * (1 - v) + b * v;
}

function smooth(v) {
  return v * v * (3 - 2 * v);
}

function randomGradient(rng) {
  const v = rng.uniform() * Math.PI * 2;
  return [Math.cos(v), Math.sin(v)];
}

// TODO: Break grad into two params
function gradient(ox, oy, grad, px, py) {
  return grad[0] * (px - ox) + grad[1] * (py - oy);
}

// Ported from the Go source: https://golang.org/src/math/rand/rand.go?s=5071:5103#L154
function seededPermutation(rng, maxExclusive) {
  const values = new Uint16Array(maxExclusive);
  for (let i = 0; i < maxExclusive; i++) {
    const j = rng.range(0, i);
    values[i] = values[j];
    values[j] = i;
  }
  return values;
}


exports.NoiseGenerator = class NoiseGenerator {
  constructor(seed) {
    this.seed = seed;
    this.rng = new rng.MT(seed);

    this.permutations = seededPermutation(this.rng, TERRAIN_PERLIN_PERIOD);

    this.rgradients = new Array(TERRAIN_PERLIN_PERIOD);
    for (let i = 0; i < TERRAIN_PERLIN_PERIOD; i++) {
      this.rgradients[i] = randomGradient(this.rng);
    }
  }

  getGradient(x, y) {
    const idx = this.permutations[x & TERRAIN_PERLIN_PERIOD_M1] + this.permutations[y & TERRAIN_PERLIN_PERIOD_M1];
    return this.rgradients[idx & TERRAIN_PERLIN_PERIOD_M1]
  }

  get2d(x, y) {
    const x0 = Math.floor(x);
    const y0 = Math.floor(y);
    const x1 = x0 + 1;
    const y1 = y0 + 1;

    const gradients0 = this.getGradient(x0, y0);
    const gradients1 = this.getGradient(x1, y0);
    const gradients2 = this.getGradient(x0, y1);
    const gradients3 = this.getGradient(x1, y1);

    const v0 = gradient(x0, y0, gradients0, x, y);
    const v1 = gradient(x0 + 1, y0, gradients1, x, y);
    const v2 = gradient(x0, y0 + 1, gradients2, x, y);
    const v3 = gradient(x0 + 1, y0 + 1, gradients3, x, y);
    const fx = smooth(x - x0);
    const vx0 = lerp(v0, v1, fx);
    const vx1 = lerp(v2, v3, fx);
    const fy = smooth(y - y0);
    return lerp(vx0, vx1, fy);
  }

  get2dNormal(x, y, max) {
    const v = this.get2d(x * PERLIN_FREQUENCY, y * PERLIN_FREQUENCY) * 0.5 + 0.5;
    return Math.pow(v, PERLIN_DILATION) * max;
  }

  getCentered2d(x, y, freq, bounds) {
    return this.get2d(x * freq, y * freq) * bounds;
  }

  fillGrid(x, y, grid, width, height, max) {
    for (let i = 0; i < height; i++) {
      for (let j = 0; j < width; j++) {
        grid[i * width + j] = (
          this.get2dNormal(x + j, y + i, max - 1) +
          this.getCentered2d(x + j, y + i, PERLIN_BIOME_FREQUENCY, max - 1) +
          PERLIN_UPLIFT
        );
      }
    }
  }
};
